Multinomial logistic regression

jolars opened the pull request:
This pull requests adds support for multnomial logistic regression. In order to accommodate for this, a rather significant number of changes have been made:

* `saga.h` now features an additional loop over the number of classes for each update,
* the Family object now contains the response (y) and should perhaps more appropriately be named Response instead now,
* gradient and predictions are now stored as vectors instead of scalars,
* plus a number of small adjustments to make everything work with the added dimension.

Some other changes also include:
* The data set permeability has been replaced with the S&P letters data set contained in `houses` because the former was a p >> n data set which our algorithm doesn't really seem tailored to.
* The lambda path is now computed and returned exactly as in glmnet, which wasn't really the case before.
* Doubled the number of tests (although this mostly involves permutations of tests across the different families and prediction types).

Here is a simple demonstration:

```r 
library(sgdnet)
x <- as.matrix(iris[, 1:4])
y <- iris[, 5]

sfit <- sgdnet(x, y, family = "multinomial")
plot(sfit)
```
![image](https://user-images.githubusercontent.com/13087841/42688052-187082a2-869b-11e8-9e7e-b2b973c3f006.png)

So far only the regular lasso penalty is supported and not the group lasso.

jolars committed:
Support multinomial family

jolars committed:
Move softmax prediction to predict.sgdnet and drop temporary and unused code

jolars committed:
Expand and fix deviance tests

jolars committed:
Rectify null deviance and lambda max computations

* Correct null deviance computations for binomial and multinomial families. In the former case, the output was wrong whenever the intercept was not fit. In the latter it was always wrong.
* Fix tests for null deviance computations, which failed because of erroneous use of testthat::expect_equal().
* Simplify lambda calculations to avoid code repetition.
* Change domain for binomial family to {0,1 }

jolars committed:
Drop unused extract_family() function

jolars committed:
Add tests of deviance(), package loading, and plotting to improve coverage

jolars committed:
Remove fixed TODO

michaelweylandt commented on R/data.R:
```diff
@@ -14,24 +14,33 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-#' Permeability
+#' S&P Letters Data
```
Is there a better name for this data set? Documenting it by the name of the journal in which it was published doesn't seem right.

jolars replied:
Good point, but I am going to switch this data set for something smaller anyway.

michaelweylandt commented on R/data.R:
```diff
@@ -14,24 +14,33 @@
+#'   \item{x}{a dense feature matrix with median house value, median income,
+#'            housing median age, total rooms, total bedrooms, population,
+#'            households, latitude, and longitude}
 #'   \item{y}{a numeric vector of permeability values for 165 compounds}
```
Should the documentation for `y` be changed also?

jolars replied:
Thanks, but as above, this is going to change anyway.

michaelweylandt commented on R/plot.sgdnet.R:
```diff
@@ -60,8 +91,11 @@ plot.sgdnet <- function(x, xvar = c("norm", "lambda", "dev"), ...) {
-           plot_data$xval <- rowSums(abs(beta))
+             expression(group("|", group("|", hat(beta), "|"), "|")[1])
+           plot_data$xval <- if (is.list(beta))
+             rowSums(sapply(beta, function(x) colSums(abs(as.matrix(x)))))
```
`vapply` is a bit more defensive (and a bit faster) than `sapply`, so I prefer it in packages.

jolars replied:
Yes, good point.

michaelweylandt commented on R/predict.sgdnet.R:
```diff
@@ -129,6 +163,69 @@ lambda_interpolate <- function(lambda, s) {
+  lambda <- object$lambda
+  which <- match(s, lambda, FALSE)
+
+  if (!all(which > 0)) {
```
Clearer to write `!all(s %in% object$lambda)`?

jolars replied:
Definitely!

michaelweylandt commented on R/predict.sgdnet.R:
```diff
@@ -129,6 +163,69 @@ lambda_interpolate <- function(lambda, s) {
+#'
+#' @return A matrix (or list of matrices) with new coefficients based
+#'   on linearly interpolating from new and old lambda values.
+#' @keywords internal
```
Should these internal functions have a `@noRd` tag? I don't think we want to expose documentation for them to our users (and I'm too lazy to check if `roxygen2` will document these by default).

jolars replied:
I think you're right... although I do sometimes think it is nice to have the documentation included but not visible because it means I can do ?function() whenever I find an internal function in the code that I don't remember the purpose of, but perhaps it is cleaner to add @noRd anyway.

michaelweylandt commented on R/utils.R:
```diff
@@ -22,7 +22,7 @@
 #'
 #' @keywords internal
 extract_family <- function(x) {
```
Might be a bit easier to implement this as an S3 generic with methods like `extract_family.sgdnet_binomial <- function(x) "binomial"`

michaelweylandt commented on tests/testthat/test-assertions.R:
```diff
@@ -0,0 +1,38 @@
+
+test_that("output types are those that we expect", {
+  expect_equal(2 * 2, 4)
+})
```
Left this in by mistake.

jolars replied:
Thanks

michaelweylandt commented on tests/testthat/test-families.R:
```diff
@@ -0,0 +1,46 @@
+context("general family tests")
```
This is nice.

michaelweylandt commented on tests/testthat/test-families.R:
```diff
@@ -0,0 +1,46 @@
+
+  grid <- expand.grid(
+    family = c("gaussian", "binomial", "multinomial"),
+    intercept = TRUE,
```
Is there a reason we aren't testing both TRUE and FALSE here? The issue in #2 only comes up when sparse=TRUE, right?

jolars replied:
One reason was that glmnet does some rather odd things when intercept = FALSE.

See this for instance:

```r
x <- as.matrix(subset(mtcars, select = c("mpg", "cyl", "disp")))
y <- mtcars$drat

gfit <- glmnet(x, y, intercept = FALSE)
gfit$lambda
#> [1] 12.539327 11.425367 10.410369  9.485541  8.642871
coef(gfit)
#> 4 x 5 sparse Matrix of class "dgCMatrix"
#>             s0         s1        s2         s3         s4
#> (Intercept)  . .          .         .          .         
#> mpg          . 0.01505858 0.0287794 0.03729123 0.04470854
#> cyl          . .          .         0.01518483 0.03029293
#> disp         . .          .         .          .         
```

the path's length is only 5 and stops even before disp enters the model. Somebody reported this on stats exchange: <https://stats.stackexchange.com/questions/243347/why-is-cv-glmnet-returning-absurd-coefficients-when-intercept-term-is-omitted/243434>

An idea is to use the lambda path from sgdnet in the call to glmnet instead, since it is tested separately anyway.

michaelweylandt commented on tests/testthat/test-lambda-path.R:
```diff
@@ -1,44 +1,153 @@
+    sfit <- do.call(sgdnet, pars)
+    gfit <- do.call(glmnet, pars)
+
+    expect_equal(sfit$lambda, gfit$lambda)
```
Is there a reason you aren't testing `coef(sfit)` and `coef(gfit)` equal here (or elsewhere)? I think it's probably useful to check that pathwise fitting gets the right result (in addition to sequential fitting) to protect against getting something wrong vs warm-starting.

jolars replied:
Since coef(fit) is the same as predict(fit, type = "coefficients") it is actually being tested for instance here:

https://github.com/jolars/sgdnet/blob/134d09d54b63bb2e925b2c919f713c5e375afa4b/tests/testthat/test-families.R#L40-L44

The purpose of `test-lambda-path.R` is only to check the calculations of the lambda max and the lambda path, not the actual fits from them, which we do in `test-families.R` and elsewhere.

michaelweylandt reviewed (approved):
I didn't read the C++ in too much detail (though I'm happy to if there's a particular spot you want feedback on) but the tests look good

jolars committed:
Fix erroneous variance computations for sparse features

jolars committed:
Use vapply instead of sapply

jolars committed:
Don't add .rd documentation for internal functions

jolars committed:
Drop autogenerated test

jolars committed:
Change variable_names to response_names here

jolars committed:
Expand with additional tests for assertions

jolars committed:
Replace current data sets with smaller ones

jolars committed:
Spread out test

jolars committed:
Handle data frames as input to newx appropriately

jolars committed:
Check that we can provide vectors and data frames as input to newx in predict.sgdnet

jolars committed:
clip unwanted parenthesis

jolars committed:
Scrap whitelines

jolars committed:
Correct urls

jolars committed:
Fix multinomial example

jolars committed:
Fix wine example

jolars committed:
Fix multinomial example

jolars committed:
Fix win example

jolars committed:
Correct erroneous computations of ColNormsMax

jolars committed:
Don't require that first solution is completely sparse when first coefficient is infinitesimal

jolars committed:
Format comment properly

jolars committed:
Update documentation

jolars committed:
Run tests over nonzero and class predictions too

jolars committed:
Compare linear interpolation against exact fit

jolars committed:
Ditch left-over code fragments

jolars committed:
Test refitting part of predict.sgdnet() outside testthat infrastructure

Currently, the exact = TRUE option to predict.sgdnet(), which invokes update() on the sgdnet fit leads to an infinite testthat loop.

jolars committed:
Drop tolerance in linear interpolation test

jolars committed:
Strip empty line

jolars committed:
Run tests for predictions over combinations of input

jolars committed:
Update test coverage
